### DISCUSSION TITLE
FFMpeg to use Nvidia Encoder Decoder?

### DISCUSSION THREAD

**Discussion by thebearot** [Motion-Project/motionplus#52] [Q&A]
Hey All,

I believe this question has been asked a while back (via Google Search) but there was no solid response or answer given. I am now raising this question again because of my project. 

On my test machine, I have a Intel 2600k + Nvidia 970 running in Ubuntu with FFMpeg compiled with cuda (encoder) and libnpp (decoder). Running this command completely offloads the work to Nvidia 970 (1% CPU Usage) - "./ffmpeg -hwaccel cuda -i input -c:v h264_nvenc -preset slow output.mkv". Without decoder, I get about 13% CPU usage per feed (rtsp) and about 20% per CPU usage without both.

From the above, I take that we can make use of FFMpeg's capability with hardware acceleration. Thus, in motinoplus' camera configurations, I added:

- netcam_high_params hwaccel=cuda,decoder=h264_nvenc
- netcam_params hwaccel=cuda,decoder=h264_nvenc

The params do not work so here I am seeking help on the forum. My aim here is to add more video cards instead of more hosts as I add more cameras. Hopefully someone who's got similar setup working can point me to the right direction.

Thanks!

-T

**Comment by Mr-Dave** [comment]
I don't have those so it would be hard to test/validate anything.  With that said, I would recommend taking a multi-phased approach.  

First, tackle the decoding of a rtsp stream.  The decoders that are available for the machine are listed in the log.  For MotionPlus, only VAAPI is available and it is in the netcam module [here](https://github.com/Motion-Project/motionplus/blob/c38c9d9e65e8d44061883b776032360402d4d9c1/src/netcam.cpp#L480) .  Someone provided a PR for Motion to provide cuda decoding [here](https://github.com/Motion-Project/motion/blob/f710f915e98f2ff001804360c09868cfac8182d6/src/netcam_rtsp.c#L377).  But, since this happened after MotionPlus broke away from Motion it didn't make it.  And since I don't have that HW, I have no method to implement and test it.  The code between Motion and MotionPlus is very similar but not identical so it should be possible for you to cut and paste with just a few revisions.    

The next phase would be encoding movies.  The preferred option of course is to just avoid it entirely and use the passthrough option....but if the user must, then I'd think that it would be using the preferred codec option as mp4:h264_nvenc

**Comment by luqasz** [comment]
How to get all available motionplus hardware decoders ?

**Reply by Mr-Dave** [reply]
They are listed in the log as the INF level.  e.g. 
```
[INF][NET][00:av02:MyCam1] netcam_hwdecoders: Norm: HW Devices: 
[INF][NET][00:av02:MyCam1] netcam_hwdecoders: Norm: vdpau (not implemented)
[INF][NET][00:av02:MyCam1] netcam_hwdecoders: Norm: cuda (available)
[INF][NET][00:av02:MyCam1] netcam_hwdecoders: Norm: vaapi (available)
[INF][NET][00:av02:MyCam1] netcam_hwdecoders: Norm: qsv (not implemented)
[INF][NET][00:av02:MyCam1] netcam_hwdecoders: Norm: drm (available)
[INF][NET][00:av02:MyCam1] netcam_hwdecoders: Norm: opencl (not implemented)
[INF][NET][00:av02:MyCam1] netcam_hwdecoders: Norm: vulkan (not implemented)
```

**Reply by luqasz** [reply]
Thx.  
```
lis 28 20:37:30 dbox motionplus[3444507]: [INF][NET][00:av02:kitajec] netcam_hwdecoders: Norm: HW Devices:
lis 28 20:37:30 dbox motionplus[3444507]: [INF][NET][00:av02:kitajec] netcam_hwdecoders: Norm: vdpau (not implemented)
lis 28 20:37:30 dbox motionplus[3444507]: [INF][NET][00:av02:kitajec] netcam_hwdecoders: Norm: cuda (not implemented)
lis 28 20:37:30 dbox motionplus[3444507]: [INF][NET][00:av02:kitajec] netcam_hwdecoders: Norm: vaapi (available)
lis 28 20:37:30 dbox motionplus[3444507]: [INF][NET][00:av02:kitajec] netcam_hwdecoders: Norm: qsv (not implemented)
lis 28 20:37:30 dbox motionplus[3444507]: [INF][NET][00:av02:kitajec] netcam_hwdecoders: Norm: drm (not implemented)
lis 28 20:37:30 dbox motionplus[3444507]: [INF][NET][00:av02:kitajec] netcam_hwdecoders: Norm: opencl (not implemented)
lis 28 20:37:30 dbox motionplus[3444507]: [INF][NET][00:av02:kitajec] netcam_hwdecoders: Norm: vulkan (not implemented)
```
Passing `decoder=vaapi` causes:
```
sty 12 11:51:02 dbox motionplus[146380]: [ERR][NET][01:av05:garaz_36] netcam_decoder_error: Norm: hwctx: Generic error in an external library
sty 12 11:51:02 dbox motionplus[146380]: [ERR][NET][01:av05:garaz_36] netcam_decoder_error: Norm: hwvaapi_init: Failed
sty 12 11:51:02 dbox motionplus[146380]: [ERR][NET][01:ml01:garaz_36] netcam_open_context: Norm: Unable to open codec context: Permission denied.
sty 12 11:51:02 dbox motionplus[146380]: [ERR][ALL][01:ml01:garaz_36] mlp_init_cam_start: Failed to start camera.
sty 12 11:51:02 dbox motionplus[146380]: [ERR][ALL][01:ml01:garaz_36] mlp_init_firstimage: Unable to open camera
sty 12 11:51:10 dbox motionplus[146380]: [WRN][ALL][01:ml01:garaz_36] mlp_retry: Retrying until successful connection with camera
```
I have `va-driver-all` on debian installed. i7-8700 cpu.

**Reply by Mr-Dave** [reply]
This may help.  What I found just recently was that for HW decoding to work, you need to run as sudo if Motionplus is running as a service or in a headless connection.  I believe that VAAPI needs access to X and it would work when running from a console but not as a service or via ssh.

**Reply by luqasz** [reply]
Could you please describe what you did exactly to enable hw decoding ?

**Reply by Mr-Dave** [reply]
For my computer, it is simply `netcam_params decoder=vaapi`

**Reply by luqasz** [reply]
Thx. Do you run it as a service via systemd ar root ? Any additional packages installed ? I am running debian.

**Reply by luqasz** [reply]
When ran as root with vaapi, motionplus segfaults.

**Reply by Mr-Dave** [reply]
That was fixed recently in the current master branch.

**Reply by luqasz** [reply]
Thx. I will build from master then.
